apps/prairielearn/src/models/enrollment.ts

```diff
@@ -158,8 +228,157 @@ export async function generateAndEnrollUsers({
+  enrollment_id,
+  agent_user_id,
+  pending_uid,
+  agent_authn_user_id,
+}: {
+  enrollment_id: string;
+  agent_user_id: string | null;
+  pending_uid: string;
+  agent_authn_user_id: string | null;
+}): Promise<Enrollment> {
+  const oldEnrollment = await selectEnrollmentById({ id: enrollment_id });
+
+  assert(oldEnrollment.status !== 'invited');
+  assert(oldEnrollment.status !== 'joined');
+
+  const newEnrollment = await queryRow(
+    sql.invite_existing_enrollment,
+    { enrollment_id, pending_uid },
+    EnrollmentSchema,
+  );
+
+  await insertAuditEvent({
+    table_name: 'enrollments',
+    action: 'update',
+    action_detail: 'invited',
+    row_id: newEnrollment.id,
+    old_row: oldEnrollment,
+    new_row: newEnrollment,
+    agent_user_id,
+    agent_authn_user_id,
+  });
+
+  return newEnrollment;
+}
+
+async function inviteNewEnrollment({
+  course_instance_id,
+  pending_uid,
+  agent_user_id,
+  agent_authn_user_id,
+}: {
+  course_instance_id: string;
+  pending_uid: string;
+  agent_user_id: string | null;
+  agent_authn_user_id: string | null;
+}) {
+  const newEnrollment = await queryRow(
+    sql.invite_new_enrollment,
+    { course_instance_id, pending_uid },
+    EnrollmentSchema,
+  );
+
+  await insertAuditEvent({
+    table_name: 'enrollments',
+    action: 'insert',
+    action_detail: 'invited',
+    row_id: newEnrollment.id,
+    new_row: newEnrollment,
+    subject_user_id: null,
+    agent_user_id,
+    agent_authn_user_id,
+  });
+
+  return newEnrollment;
+}
+/**
+ * Invite a student by uid. This does not prevent being invited via
+ * If there is an existing enrollment with the given uid, it will be updated to a invitation.
+ * If there is no existing enrollment, a new enrollment will be created.
+ */
+export async function inviteStudentByUid({
+  course_instance_id,
+  uid,
+  agent_user_id,
+  agent_authn_user_id,
+}: {
+  course_instance_id: string;
+  uid: string;
+  agent_user_id: string | null;
+  agent_authn_user_id: string | null;
+}): Promise<Enrollment> {
+  return await runInTransactionAsync(async () => {
+    const existingEnrollment = await selectOptionalEnrollmentByUid({
+      course_instance_id,
+      uid,
+    });
+
+    if (existingEnrollment) {
+      if (existingEnrollment.user_id) {
+        await selectAndLockUserById(existingEnrollment.user_id);
+      }
+      await selectAndLockEnrollmentById(existingEnrollment.id);
+      return await dangerouslyInviteExistingEnrollment({
+        enrollment_id: existingEnrollment.id,
+        agent_user_id,
+        pending_uid: uid,
+        agent_authn_user_id,
+      });
+    }
+
+    return await inviteNewEnrollment({
+      course_instance_id,
+      pending_uid: uid,
+      agent_user_id,
+      agent_authn_user_id,
+    });
+  });
+}
+
+export async function selectAndLockEnrollmentById(id: string) {
+  await queryRow(sql.select_and_lock_enrollment_by_id, { id }, EnrollmentSchema);
+}
```

Comment:
`select...` in a name implies that it will return something.

```suggestion
export async function selectAndLockEnrollmentById(id: string) {
  return await queryRow(sql.select_and_lock_enrollment_by_id, { id }, EnrollmentSchema);
}
```